docs/design_proposals/multi_environment_proposal.md

```diff
@@ -0,0 +1,277 @@
+5. **Fixed Lockfiles**: It's crucial to preserve fixed lockfiles for consistency and predictability. Solutions must ensure reliability not just for authors but also for end-users, particularly at the time of lockfile creation.
+
+## Proposed Solution
+!!! important
+    This is a proposal, not a final design. The proposal is open for discussion and will be updated based on the feedback.
+
+### Feature & Environment Set Definitions
+Introduce environment sets into the `pixi.toml` this describes environments based on `feature`'s. Introduce features into the `pixi.toml` that can describe parts of environments.
+As an environment goes beyond just `dependencies` the `features` should be described including the following fields:
+
+- `dependencies`: The conda package dependencies
+- `pypi-dependencies`: The pypi package dependencies
+- `system-requirements`: The system requirements of the environment
+- `activation`: The activation information for the environment
+- `platforms`: The platforms the environment can be run on.
+- `channels`: The channels used to create the environment.
+- `target`: All the above features but also separated by targets.
+- `tasks`: Feature specific tasks, tasks in one environment are selected as default tasks for the environment.
+
+
+```toml title="Default features" linenums="1"
+[dependencies] # short for [feature.default.dependencies]
+python = "*"
+numpy = "==2.3"
+
+[pypi-dependencies] # short for [feature.default.pypi-dependencies]
+pandas = "*"
+
+[system-requirements] # short for [feature.default.system-requirements]
+libc = "2.33"
+
+[activation] # short for [feature.default.activation]
+scripts = ["activate.sh"]
+```
+
+```toml title="Different dependencies per feature" linenums="1"
+[feature.py39.dependencies]
+python = "~=3.9.0"
+[feature.py310.dependencies]
+python = "~=3.10.0"
+[feature.test.dependencies]
+pytest = "*"
+```
+
+```toml title="Full set of environment modification in one feature" linenums="1"
+[feature.cuda]
+dependencies = {cuda = "x.y.z", cudnn = "12.0"}
+pypi-dependencies = {torch = "1.9.0"}
+platforms = ["linux-64", "osx-arm64"]
+activation = {scripts = ["cuda_activation.sh"]}
+channels = ["nvidia"] # Would concat instead of overwrite, so the default channels are still used.
+tasks = { warmup = "python warmup.py" }
+target.osx-arm64 = {dependencies = {mlx = "x.y.z"}}
+```
+
+
+```toml title="Define tasks as defaults of an environment" linenums="1"
+[feature.test.tasks]
+test = "pytest"
+
+[environments]
+test = ["test"]
+
+# `pixi run test` == `pixi run --envrionment test test`
+```
+
+The environment definition should contain the following fields:
+
+- `features: Vec<Feature>`: The features that are included in the environment set, which is also the default field in the environments.
+- `environments: Vec<Environment>`: The environments that are included in the environment set. When environments is used, the extra features are **on top** of the included environments.
```

Comment:
Now that I think about it, I'm not sure whether we should really do `Vec<Environment>` instead of `Environment`.
Let's look at this example

```toml
[environments]
"foo" = ["py39", "test1"]
"bar" = ["py39", "test2"]
"xyz" = {environments = ["foo", "bar"]}
```

First, pixi solves `foo` and `bar` separately. For some reason, `foo` gets the package `python=3.9.11` and `bar` gets the package `3.9.12`.
Then, `xyz` will not be solvable anymore since it has the constraints `3.9.11` and `3.9.12`.

We could mitigate this by doing one large solve: first solve `xyz` with _all_ requirements of `xyz` and its parents combined and then create `foo` and `bar` by selecting a strict subset of `xyz`. This will have the consequence that `foo` will be solved differently depending if it is a "parent" of another environment or not (which I personally would find a bit annoying).

I also cannot think of a proper use case where you would need to explicitly pin two parent environments.
My "prod" use case is completely covered by having only one environment as a parent.

Reply:
Yeah I was also doubting it a bit. Also about the name, we shortly mentioned `extends = "foo"` in the call. Maybe that could be a better name and then keep it to a `extends: Environment` instead of `Vec<>`. This might also work with the future `stack_on = "env"` from wolf. 

Maybe this could work instead of `environments`:
- `extends-locked`: Single environment which gets extended with some more features but keeps the original locked dependency set.
- `extends` : Single environment which gets extended with some more features and resolves the merged dependency set.

Reply:
Do you have a use case for the second option?
Seems like this is semantically equivalent to just creating a second env with the additional features in the `environment` section. Not sure how I feel about creating two ways to do basically the same thing.

Reply:
Yeah you are right, it would just be a convenience function. I have a hard time finding an intuitive name for that first field so I think I was making it right be giving the second option too.